core/modules/replica/HttpCatalogsModule.cc

```diff
@@ -0,0 +1,198 @@
+
+    sendData(resp, _catalogsReport);
+}
+
+
+json HttpCatalogsModule::_databaseStats(string const& database, bool dummyReport) const {
+
+    auto const config = controller()->serviceProvider()->config();
+    auto const databaseServices = controller()->serviceProvider()->databaseServices();
+
+    vector<unsigned int> chunks;
+    if (not dummyReport) databaseServices->findDatabaseChunks(chunks, database);
+
+    vector<ReplicaInfo> replicas;
+    if (not dummyReport) databaseServices->findDatabaseReplicas(replicas, database);
+
+    json result;
+    result["chunks"]["unique"] = chunks.size();
+    result["chunks"]["with_replicas"] = replicas.size();
+
+    map<string, map<string, size_t>> stats;
+    for (auto&& table: config->databaseInfo(database).partitionedTables) {
+        stats[table]["data_unique_in_chunks_data"] = 0;
+        stats[table]["data_unique_in_chunks_index"] = 0;
+        stats[table]["data_unique_in_overlaps_data"] = 0;
+        stats[table]["data_unique_in_overlaps_index"] = 0;
+        stats[table]["data_with_replicas_in_chunks_data"] = 0;
+        stats[table]["data_with_replicas_in_chunks_index"] = 0;
+        stats[table]["data_with_replicas_in_overlaps_data"] = 0;
+        stats[table]["data_with_replicas_in_overlaps_index"] = 0;
+    }
+    set<unsigned int> uniqueChunks;
+    for (auto&& replica: replicas) {
+        bool isUniqueChunk = false;
+        if (uniqueChunks.count(replica.chunk()) == 0) {
+            uniqueChunks.insert(replica.chunk());
+            isUniqueChunk = true;
+        }
+        for (auto&& f: replica.fileInfo()) {
+            auto const table     = f.baseTable();
+            auto const isData    = f.isData();
+            auto const isIndex   = f.isIndex();
+            auto const isOverlap = f.isOverlap();
+            auto const size      = f.size;
+            if (isUniqueChunk) {
+                if (isData) {
+                    if (isOverlap) stats[table]["data_unique_in_overlaps_data"] += size;
+                    else           stats[table]["data_unique_in_chunks_data"]   += size;
+                }
+                if (isIndex) {
+                    if (isOverlap) stats[table]["data_unique_in_overlaps_index"] += size;
+                    else           stats[table]["data_unique_in_chunks_index"]   += size;
+                }
+            }
+            if (isData) {
+                if (isOverlap) stats[table]["data_with_replica_in_overlaps_data"] += size;
+                else           stats[table]["data_with_replica_in_chunks_data"]   += size;
+            }
+            if (isIndex) {
+                if (isOverlap) stats[table]["data_with_replica_in_overlaps_index"] += size;
+                else           stats[table]["data_with_replica_in_chunks_index"]   += size;
+            }            
+        }
+    }
+    for (auto&& table: config->databaseInfo(database).partitionedTables) {
+        result["tables"][table]["is_partitioned"] = 1;
```

Comment:
can you cache a reference to a json object at `result["tables"][table]`?

Reply:
Good point! Refactored.
Though, to be honest, caching references makes me feel a bit uneasy because what I'm dealing with is not a Standard Library's container. This is a class from the `nlohmann` JSON library which may not behave the way you might expect it. It may reallocate entries on each insert/modifications. I have already seen a few cases of the very strange (counter-intuitive) behavior in tis library. So, if my refactoring will break the code then all blame on you 😹